dlt/sources/helpers/rest_client/paginators.py

```diff
@@ -7,45 +7,71 @@
 
 
 class BasePaginator(ABC):
+    """
+    Base class for all paginator implementations. Paginators are used
+    to handle paginated responses from RESTful APIs.
+
+    See `RESTClient.paginate()` for example usage.
+    """
+
     def __init__(self) -> None:
         self._has_next_page = True
         self._next_reference: Optional[str] = None
 
     @property
     def has_next_page(self) -> bool:
-        """
-        Check if there is a next page available.
+        """Determines if there is a next page available.
 
         Returns:
-            bool: True if there is a next page available, False otherwise.
+            bool: True if a next page is available, otherwise False.
         """
         return self._has_next_page
 
     @property
     def next_reference(self) -> Optional[str]:
+        """The reference to the next page, such as a URL or a cursor.
+
+        Returns:
+            Optional[str]: The reference to the next page if available,
+                otherwise None.
+        """
         return self._next_reference
 
     @next_reference.setter
     def next_reference(self, value: Optional[str]) -> None:
+        """Sets the reference to the next page and updates the availability
+        of the next page.
+
+        Args:
+            value (Optional[str]): The reference to the next page.
+        """
         self._next_reference = value
         self._has_next_page = value is not None
 
     @abstractmethod
     def update_state(self, response: Response) -> None:
-        """Update the paginator state based on the response.
+        """Updates the paginator's state based on the response from the API.
+
+        This method should extract necessary pagination details (like next page
+        references) from the response and update the paginator's state
+        accordingly.
 
         Args:
-            response (Response): The response object from the API.
+            response (Response): The response object from the API request.
```

Comment:
This method must call `next_reference` setter with a value of next reference right? otherwise `has_next_page` is not going to work correctly. Maybe we should mention this here?

I also wonder if we really need to force `next_referece` in base class and just leave `has_next_page` abstract? This property is kind of useful in  `BaseNextUrlPaginator` (and I wonder why it is public?)

Reply:
otherwise this is really high quality!

Reply:
> This method must call `next_reference` setter with a value of next reference right? otherwise `has_next_page` is not going to work correctly. Maybe we should mention this here?
> 
> I also wonder if we really need to force `next_referece` in base class and just leave `has_next_page` abstract? This property is kind of useful in `BaseNextUrlPaginator` (and I wonder why it is public?)

Yes, both are valid points: 
1. `next_reference` is useless in offset/limit pagination (or page-based), so having it in the base class is not a good design. 
Let's address it a follow-up PR. 
2. `update_state` needs to modify `_has_next_page` indeed. I'll add this to the docstring.